Use crypto.Signer whenever possible

hoihochan opened the pull request:
For certificate-based authentication, it is possible that the private key resides in hardware such as Trusted Platform Module (TPM) and the only way in Golang to access it via the crypto.Signer interface.

Any code paths that deal with a private key should use the crypto.Signer interface which comes with a function called Public() and it can be used to determine the type of key used.

Sean-Der commented:
LGTM @hoihochan!

Would it be possible to fix the API breaks? Maybe leave the old functions?

```
github.com/pion/dtls/v3/pkg/crypto/selfsign
  Incompatible changes:
  - SelfSign: changed from func(crypto.PrivateKey) (crypto/tls.Certificate, error) to func(crypto.Signer) (crypto/tls.Certificate, error)
  - WithDNS: changed from func(crypto.PrivateKey, string, ...string) (crypto/tls.Certificate, error) to func(crypto.Signer, string, ...string) (crypto/tls.Certificate, error)

github.com/pion/dtls/v3/pkg/crypto/signaturehash
  Incompatible changes:
  - SelectSignatureScheme: changed from func([]Algorithm, crypto.PrivateKey) (Algorithm, error) to func([]Algorithm, crypto.Signer) (Algorithm, error)'
+ GOAPIDIFF_RC=1
+ set -e
++ dd if=/dev/urandom bs=15 count=1 status=none
++ base64
+ EOF=91zKzIfpY5R4+8vuyenb
+ echo 'output<<91zKzIfpY5R4+8vuyenb'
+ echo '
github.com/pion/dtls/v3/pkg/crypto/selfsign
  Incompatible changes:
  - SelfSign: changed from func(crypto.PrivateKey) (crypto/tls.Certificate, error) to func(crypto.Signer) (crypto/tls.Certificate, error)
  - WithDNS: changed from func(crypto.PrivateKey, string, ...string) (crypto/tls.Certificate, error) to func(crypto.Signer, string, ...string) (crypto/tls.Certificate, error)

github.com/pion/dtls/v3/pkg/crypto/signaturehash
  Incompatible changes:
  - SelectSignatureScheme: changed from func([]Algorithm, crypto.PrivateKey) (Algorithm, error) to func([]Algorithm, crypto.Signer) (Algorithm, error)'
  ```

hoihochan commented:
> LGTM @hoihochan!
> 
> Would it be possible to fix the API breaks? Maybe leave the old functions?
> 
> ```
> github.com/pion/dtls/v3/pkg/crypto/selfsign
>   Incompatible changes:
>   - SelfSign: changed from func(crypto.PrivateKey) (crypto/tls.Certificate, error) to func(crypto.Signer) (crypto/tls.Certificate, error)
>   - WithDNS: changed from func(crypto.PrivateKey, string, ...string) (crypto/tls.Certificate, error) to func(crypto.Signer, string, ...string) (crypto/tls.Certificate, error)
> 
> github.com/pion/dtls/v3/pkg/crypto/signaturehash
>   Incompatible changes:
>   - SelectSignatureScheme: changed from func([]Algorithm, crypto.PrivateKey) (Algorithm, error) to func([]Algorithm, crypto.Signer) (Algorithm, error)'
> + GOAPIDIFF_RC=1
> + set -e
> ++ dd if=/dev/urandom bs=15 count=1 status=none
> ++ base64
> + EOF=91zKzIfpY5R4+8vuyenb
> + echo 'output<<91zKzIfpY5R4+8vuyenb'
> + echo '
> github.com/pion/dtls/v3/pkg/crypto/selfsign
>   Incompatible changes:
>   - SelfSign: changed from func(crypto.PrivateKey) (crypto/tls.Certificate, error) to func(crypto.Signer) (crypto/tls.Certificate, error)
>   - WithDNS: changed from func(crypto.PrivateKey, string, ...string) (crypto/tls.Certificate, error) to func(crypto.Signer, string, ...string) (crypto/tls.Certificate, error)
> 
> github.com/pion/dtls/v3/pkg/crypto/signaturehash
>   Incompatible changes:
>   - SelectSignatureScheme: changed from func([]Algorithm, crypto.PrivateKey) (Algorithm, error) to func([]Algorithm, crypto.Signer) (Algorithm, error)'
> ```

I could, but they look like internal functions that is not shared in the public documentation so let me know.

Sean-Der commented:
Someone could be using them. I believe `pion/webrtc` uses them to create certificates?

Can't really know for sure. We can't break API without a major bump. I think we can avoid that here though!

hoihochan commented:
@Sean-Der, I have re-based the PR and fixed a few linter issues on existing code - would appreciate if this can be merged.

kevmo314 commented on pkg/crypto/signaturehash/signaturehash.go:
```diff
@@ -40,8 +40,12 @@ func Algorithms() []Algorithm {
 
 // SelectSignatureScheme returns most preferred and compatible scheme.
 func SelectSignatureScheme(sigs []Algorithm, privateKey crypto.PrivateKey) (Algorithm, error) {
+	signer, ok := privateKey.(crypto.Signer)
```
Is it possible to still accept a `crypto.PrivateKey`? The documentation suggests that the stdlib `crypto.PrivateKey` implementations satisfy `crypto.Signer` but that's not necessarily true for custom implementations. I think this would be a breaking API change if someone provides a custom implementation that does not satisfy `crypto.Signer`?

hoihochan replied:
Either it should fail because everyone should be implementing `crypto.Signer` underneath - the `crypto.PrivateKey` was strictly there for backwards compatibility purposes.

hoihochan replied:
A custom implementation written currently would never work in the first place, because ultimately pion/dtls would invoke Go's standard crypto APIs that assumes `crypto.PrivateKey` satisfies `crypto.Signer`.
